LanguageFeatures/Patterns/type_inference_A26_t03.dart

```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion
+/// To type check a pattern p being matched against a value of type M:
+/// ...
+/// Identifier:
+///
+/// In an assignment context, the required type of p is the (unpromoted) static
+/// type of the variable that p resolves to.
+///
+/// In a matching context, the name refers to a constant. Type check the
+/// constant identifier expression in context type M.
+///
+/// In a declaration context, the required type of p is M, as is the static type
+/// of the variable introduced by p.
+///
+/// @description Check that in a declaration context, the required type of p is
+/// M, as is the static type of the variable introduced by p.
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns,records
+
+import "../../Utils/static_type_helper.dart";
+
+main() {
+  num a = 0;
```

Comment:
OK, I can see how this can happen: With a context type of `_` it can be very difficult to verify that we have `_` and not `Object?`:

```dart
class A<X> {
  const A();
}

X whatever() => throw 0;

void main() {
  switch (whatever()) { // Context type schema should be `_`.
    case const A(): ...; // There is no expression that denotes the constant A, how can we test its type argument?
    case 1: ... // Doesn't care about the context type schema.
    case {"key": 5}: ... // Can't distinguish `Object?` and `_`.
    ... // And so on.
  }
}
```

In the case where we really can't check some specific property directly we may have to check it indirectly, and there are cases where we can do this (e.g., we can check the actual type arguments of a list pattern because it will be used as the declared type of a `var x` pattern field).

So we can't expect to get rid of all redundant test cases.

But it would still be a good idea to try to avoid redundancy whenever possible, especially in the cases where the given test case doesn't actually test precisely the topic of the given test.

(Like a run-time semantics test occurring in a test PR named 'Type checking' ;-)